### ISSUE TITLE
AutoPilot fails after reverting to launch

### ISSUE THREAD

**Issue by chippydip** [krpc/krpc#220]
This might be related to #210.

After reverting to launch, AutoPilot steering no longer works and the KSP logs are spammed with this error:

```
NullReferenceException
  at (wrapper managed-to-native) UnityEngine.Component:get_rigidbody ()
  at KRPC.SpaceCenter.Utils.RotationRateController.get_Error () [0x00000] in <filename unknown>:0 
  at KRPC.SpaceCenter.Utils.RotationRateController.Update (KRPC.SpaceCenter.ControlInputs state) [0x00000] in <filename unknown>:0 
  at KRPC.SpaceCenter.Services.AutoPilot.DoAutoPiloting (KRPC.SpaceCenter.ControlInputs state) [0x00000] in <filename unknown>:0 
  at KRPC.SpaceCenter.Services.AutoPilot.Fly (.Vessel vessel, KRPC.SpaceCenter.ControlInputs state) [0x00000] in <filename unknown>:0 
  at KRPC.SpaceCenter.PilotAddon.OnFlyByWire (.Vessel vessel, .FlightCtrlState state) [0x00000] in <filename unknown>:0 
  at KRPC.SpaceCenter.PilotAddon+<Fly>c__AnonStorey0.<>m__0 (.FlightCtrlState s) [0x00000] in <filename unknown>:0 
  at (wrapper delegate-invoke) FlightInputCallback:invoke_void__this___FlightCtrlState (FlightCtrlState)
  at Vessel.FeedInputFeed () [0x00000] in <filename unknown>:0 
  at FlightInputHandler.FixedUpdate () [0x00000] in <filename unknown>:0 
```

Reverting to the VAB and then re-launching seems to fix the issue until the next revert to launch.

**Comment by djungelorm** [comment]
Yep, looks like the same issue as #210. I missed a Vessel object reference in RotationRateController.

**Comment by djungelorm** [comment]
Here's a patched version of KRPC.SpaceCenter.dll that should fix the issue: https://krpc.s3.amazonaws.com/deploy/bug/220/193.1/GameData/kRPC/KRPC.SpaceCenter.dll

Let me know if this resolves your issue.

**Comment by chippydip** [comment]
Yup, that seems to have fixed it!

Thanks for the quick response and for making such a great mod :)

**Comment by djungelorm** [comment]
Awesome. No problemo :)

**Comment by chippydip** [comment]
Now that revert to launch is working I'm noticing that the autopilot's reference frame doesn't get reset, which leads to some unexpected steering straight off the pad, but it's pretty easy to work around by just setting with one extra line of code so not a big deal:
`vessel.AutoPilot.ReferenceFrame = vessel.SurfaceReferenceFrame;`

**Comment by djungelorm** [comment]
Reopening. Ideally you shouldn't need to reset the reference frame.

Are you using vessel.SurfaceReferenceFrame before you revert to launch?

**Comment by chippydip** [comment]
I'm basically using the example launch to orbit script (translated to C#). When executing the maneuver node to circularize the script changes the reference frame:
```
            vessel.AutoPilot.ReferenceFrame = node.ReferenceFrame;
            vessel.AutoPilot.TargetDirection = Tuple.Create(0.0, 1.0, 0.0);
```
After that point, if I close my program, revert to launch, and re-run it I end up turning north to point at the horizon almost immediately. Presumably this is the leftover "up" direction from the node's reference frame, which is targeted when I set pitch/heading to 90/90.

**Comment by djungelorm** [comment]
Ah I see. Reverting to launch doesn't revert any of kRPC's state. So yes the reference frame will have been left as node.ReferenceFrame.

I'm not sure if this is a feature or a bug... Maybe we should revert the reference frame to the previous state, but then there is no "default" reference frame for the autopilot. Maybe we should require it to be set whenever the autopilot is engaged.

Which makes me think, I don't think the autopilot will disengage when you revert to launch. It probably should. (Haven't tested this though)

**Comment by chippydip** [comment]
It seemed like it disengaged when reverting, perhaps because the Vessel object changed so the FlyByWire callback was no longer registered.

Adding a ReferenceFrame parameter to Engage wouldn't be a bad idea, and it could even default to vessel.SurfaceReferenceFrame. Alternately, maybe just pointing out in the docs and examples that ReferenceFrame should be set before calling Engage would be enough.

**Comment by djungelorm** [comment]
Might be fixed by #248

**Comment by djungelorm** [comment]
#882
